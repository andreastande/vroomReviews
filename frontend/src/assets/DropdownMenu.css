.dropdownButtonWrapper {
  position: relative;
  display: flex;
  font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
}

.dropdownButton {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  gap: 1vh;
  background-color: #f4f9fe;
  border: 1px solid transparent;
  box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.25);
  font-size: 2vh;
  border-radius: 1vh;
  padding: 1.5vh;
  padding-left: 2.5vh;
  padding-right: 2.5vh;
  color: #54595e;
  font-weight: 500;
  font-family: inherit;
}

.dropdownButton:hover {
  cursor: pointer;
  box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.25);
}

.DdBlabel {
  cursor: pointer;
}

.optionsWrapper {
  display: flex;
  flex-wrap: wrap;
  gap: 0.25vh;
}

.dropdownButtonInsideWrapper {
  display: flex;
  align-items: center;
  width: calc(50% - 0.25vh);
}

.dropdownButtonInside {
  background-color: #f4f9fe;
  font-size: 2vh;
  border-radius: 0.5vh;
  padding: 1vh;
  color: #54595e;
  width: 100%;
  text-align: center;
  border: 1px solid transparent;
  font-weight: 500;
  font-family: inherit;
}

.dropdownButtonInside:hover {
  cursor: pointer;
  background-color: #e3e3e3;
}

.arrowWrapper {
  display: flex;
  width: 100%;
}

.arrowButtonWrapper {
  display: flex;
  align-items: center;
  width: calc(50% - 0.25vh);
}

.arrowButton {
  background-color: #f4f9fe;
  font-size: 2vh;
  border-radius: 0.5vh;
  padding: 1vh;
  color: #0062c4;
  width: 100%;
  text-align: center;
  border: 1px solid transparent;
  font-weight: 500;
  font-family: inherit;
  transition: transform 0.5s ease;
}

.arrowButton:hover {
  cursor: pointer;
}

.arrowButtonLeft:hover {
  transform: translateX(-10px);
}

.arrowButtonRight:hover {
  transform: translateX(10px);
}

.dropdownArrow {
  border: solid #abb5be;
  border-width: 0 2px 2px 0;
  display: inline-block;
  padding: 3px;
  transform: rotate(45deg);
  -webkit-transform: rotate(45deg);
}

.arrowLeft {
  border: solid #0062c4;
  border-width: 0 2px 2px 0;
  display: inline-block;
  padding: 3px;
  transform: rotate(45deg);
  -webkit-transform: rotate(135deg);
}

.arrowRight {
  border: solid #0062c4;
  border-width: 0 2px 2px 0;
  display: inline-block;
  padding: 3px;
  transform: rotate(45deg);
  -webkit-transform: rotate(-45deg);
}

.dropdown {
  display: flex;
  flex-wrap: wrap;
  position: absolute;
  top: 100%;
  border-radius: 10px;
  margin-top: 5px;
  z-index: 1;
  padding: 0.25rem;
  background-color: #f4f9fe;
  border: 1px solid #d8dde2;
  box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.25);
  color: #54595e;
  width: 240px;
  /* Next two lines are important for the cool effect*/
  overflow: hidden;
  transition:
    max-height 0.3s ease,
    opacity 0.3s ease;
  /* Fjern de to linjene under dersom dropdown ikke skal v√¶re sentrert i forhold til dropdownButton */
  left: 50%;
  transform: translateX(-50%);
}

.dropdown.active {
  max-height: 500px;
  opacity: 1;
}

.dropdown.closed {
  max-height: 0;
  opacity: 0;
  pointer-events: none;
}

@media (max-width: 450px) {
  .dropdown {
    width: 180px;
  }

  .dropdownButtonInside {
    padding: 0.75vh 0.25vh 0.75vh 0.25vh;
    font-size: 1.75vh;
    word-break: break-all;
  }

  .dropdownButtonInsideWrapper {
    margin-top: 0.25vh;
    margin-bottom: 0.25vh;
  }
}

.clearButton {
  margin-left: 1rem;
  background-color: #202020;
  border: none;
}

.clearButton img {
  width: 20px;
}

.clearButton:hover {
  cursor: pointer;
}
