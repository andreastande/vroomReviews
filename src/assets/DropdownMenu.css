.dropdownButtonWrapper {
    position: relative;
    display: inline-block;
}

.dropdownButton {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    gap: 1vh;
    background-color: #f4f9fe;
    border: 1px solid #d8dde2;
    box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.25);
    font-size: 2vh;
    border-radius: 1vh;
    padding: 2vh;
    padding-left: 3vh;
    padding-right: 3vh;
    color: #54595e;
}

.dropdownButton:hover {
    cursor: pointer;
    box-shadow: 0px 2px 4px 0px rgba(0, 0, 0, 0.25);
}

.DdBlabel {
    cursor: pointer;
}

.dropdownButtonInside {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    gap: 1vh;
    background-color: #f4f9fe;
    font-size: 2vh;
    border-radius: 0.5vh;
    padding: 1vh;
    padding-left: 3vh;
    padding-right: 3vh;
    color: #54595e;
    margin: 2vh, 0, 2vh, 0;
    width: 100%;
    text-align: center;
}

.dropdownButtonInside:hover {
    cursor: pointer;
    background-color: #e3e3e3;
}


.dropdownArrow {
    border: solid #abb5be;
    border-width: 0 2px 2px 0;
    display: inline-block;
    padding: 3px;
    transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
}

.dropdown {
    position: absolute;
    top: 100%;
    width: 100%;
    border-radius: 10px;
    margin-top: 5px;
    z-index: 1;
    padding: 0.25rem;
    background-color: #f4f9fe;
    border: 1px solid #d8dde2;
    box-shadow: 0px 4px 4px 0px rgba(0, 0, 0, 0.25);
    color: #54595e;
    width: max-content;
    /* Next two lines are important for the cool effect*/
    overflow: hidden;
    transition: max-height 0.3s ease, opacity 0.3s ease;
    /* Fjern de to linjene under dersom dropdown ikke skal v√¶re sentrert i forhold til dropdownButton */
    left: 50%;
    transform: translateX(-50%);
}

.dropdown.active {
    max-height: 250px;
    opacity: 1;
}

.dropdown.closed {
    max-height: 0;
    opacity: 0;
}