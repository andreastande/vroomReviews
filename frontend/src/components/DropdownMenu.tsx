import { useState, useEffect } from 'react';
import '../assets/DropdownMenu.css';
import ClearIcon from '../assets/images/clear-icon.png'

type DropdownProps = {
  filter: string;
  options: string[];
  isOpen: boolean;
  toggleDropdown: () => void;
  onSelect: (option: string, initialLoad: boolean) => void;
};

type ButtonProps = {
  name: string;
  onClick: () => void;
  checkedOption: string;
};

function ButtonInside({ name, onClick, checkedOption }: ButtonProps) {
  return (
    <div className="dropdownButtonInsideWrapper">
      <button className="dropdownButtonInside" onClick={onClick}>
        <span style={{ color: name == checkedOption ? '#00CC00' : '' }}>{name}</span>
      </button>
    </div>
  );
}

function DropdownMenu({
  filter,
  options,
  isOpen,
  toggleDropdown,
  onSelect,
}: DropdownProps) {
  const [checkedOption, setCheckedOption] = useState(filter);
  const [initialLoad, setInitialLoad] = useState(true);
  const [filterApplied, setFilterApplied] = useState(false);
  const [optionsCounter, setOptionsCounter] = useState(10)

  // Set checked option to the option that was selected last time, if there is one, otherwise set it to "All"
  useEffect(() => {
    if (initialLoad) {
      const storedFilterOption = sessionStorage.getItem(filter);
      if (storedFilterOption !== null) {
        if (storedFilterOption !== 'All') {
          setCheckedOption(storedFilterOption);
          onSelect(storedFilterOption, true);
          setFilterApplied(true)
        } else {
          setCheckedOption(filter);
          onSelect(storedFilterOption, true);
          setFilterApplied(false)
        }
      }
      setInitialLoad(false);
    } else {
      if (checkedOption == filter) {
        sessionStorage.setItem(filter, 'All');
      } else {
        sessionStorage.setItem(filter, checkedOption);
      }
    }
  }, [onSelect, checkedOption, filter, initialLoad]);

  // Set checked option to the selected option and close the dropdown
  const handleOptionClick = (option: string) => {
    if (option !== checkedOption) {
      setFilterApplied(true)
      setCheckedOption(option);
      onSelect(option, false);
    }
    toggleDropdown();
  };

  const handleDropdown = () => {
    toggleDropdown();
  };

  const handleClear = () => {
    setFilterApplied(false)
    onSelect("All", false)
    setCheckedOption(filter)
    if (isOpen) {
      toggleDropdown()
    }
  }

  return (
    <div className="dropdownButtonWrapper">
      <button className="dropdownButton" onClick={handleDropdown}>
        <label className="DdBlabel">{checkedOption}</label>
        <i className="dropdownArrow"></i>
      </button>
      <div className={`dropdown ${isOpen ? 'active' : 'closed'}`}>
        <div className="optionsWrapper">
          {options.slice(optionsCounter - 10, optionsCounter).map((option) => {
            return (
              <ButtonInside
                key={option}
                name={option}
                checkedOption={checkedOption}
                onClick={() => handleOptionClick(option)}
              />
            );
          })}
        </div>
        {optionsCounter > 10 && (
          <div className="arrowButtonWrapper">
            <button className="arrowButton" onClick={() => setOptionsCounter((prevCounter) => prevCounter - 10)}>
              Left
            </button>
          </div>
        )}
        {optionsCounter < options.length && (
          <div className="arrowButtonWrapper">
            <button className="arrowButton" onClick={() => setOptionsCounter((prevCounter) => prevCounter + 10)}>
              Right
            </button>
          </div>
        )}
      </div>
      {filterApplied && (
        <button className="clearButton" onClick={handleClear}>
          <img src={ClearIcon}></img>
        </button>
      )}
    </div>
  );
}

export default DropdownMenu;
