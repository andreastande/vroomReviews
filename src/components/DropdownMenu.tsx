import { useState, useEffect } from 'react';
import '../assets/DropdownMenu.css';

type DropdownProps = {
  filter: string;
  options: string[];
  isOpen: boolean;
  toggleDropdown: () => void;
};

type ButtonProps = {
  name: string;
  onClick: () => void;
};

function ButtonInside({ name, onClick }: ButtonProps) {
  return (
    <button className="dropdownButtonInside" onClick={onClick}>
      {name}
    </button>
  );
}

function DropdownMenu({
  filter,
  options,
  isOpen,
  toggleDropdown,
}: DropdownProps) {
  const [checkedOption, setCheckedOption] = useState(filter);

  useEffect(() => {
    localStorage.setItem(filter, checkedOption);
  }, [filter, checkedOption]);

  const handleOptionClick = (option: string) => {
    if (option === 'Alle') {
      setCheckedOption(filter);
    } else {
      setCheckedOption(option);
    }

    toggleDropdown();
  };

  const handleDropdown = () => {
    toggleDropdown();
  };

  return (
    <div className="dropdownButtonWrapper">
      <button className="dropdownButton" onClick={handleDropdown}>
        <label className="DdBlabel">{checkedOption}</label>
        <i className="dropdownArrow"></i>
      </button>
      <div className={`dropdown ${isOpen ? 'active' : 'closed'}`}>
        {options.map((option) => {
          if (option !== 'Alle' || checkedOption !== filter) {
            // only show option "Alle" when the user has applied a filter
            return (
              <ButtonInside
                key={option}
                name={option}
                onClick={() => handleOptionClick(option)}
              />
            );
          }
        })}
      </div>
    </div>
  );
}

export default DropdownMenu;