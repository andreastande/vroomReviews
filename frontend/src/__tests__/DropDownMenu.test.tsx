import { vi } from 'vitest';
import { render, cleanup, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import DropdownMenu from '../components/DropdownMenu'; // Update the path accordingly

// Mock data for testing
const mockProps = {
  filter: 'Category',
  options: ['Option1', 'Option2', 'Option3'],
  isOpen: false,
  toggleDropdown: () => {},
  onSelect: () => {},
};

// A utility function to create a mock sessionStorage for testing
const mockSessionStorage = () => {
  const sessionStorageMock = {
    getItem: vi.fn(),
    setItem: vi.fn(),
  };
  Object.defineProperty(global, 'sessionStorage', { value: sessionStorageMock });
};

// A utility function to cleanup after each test
const cleanupAfterTest = () => {
  cleanup();
  vi.restoreAllMocks();
};

describe('DropdownMenu Component', () => {
  beforeEach(() => {
    mockSessionStorage();
  });

  afterEach(cleanupAfterTest);

  test('renders without errors', () => {
    render(<DropdownMenu {...mockProps} />);
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <DropdownMenu {...mockProps} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('handles option click', () => {
    const onSelectMock = vi.fn();
    const toggleDropdownMock = vi.fn();

    const { getByText } = render(
      <DropdownMenu
        {...mockProps}
        onSelect={onSelectMock}
        toggleDropdown={toggleDropdownMock}
      />
    );


    const option1 = getByText('Option1');
    userEvent.click(option1);

    // Assertions
    expect(option1).toBeTruthy();
    expect(onSelectMock).toHaveBeenCalled();
  });

  test('handles dropdown toggle', () => {
    const onSelectMock = vi.fn();
    const toggleDropdownMock = vi.fn();

    const { getByTestId } = render(
      <DropdownMenu
        {...mockProps}
        onSelect={onSelectMock}
        toggleDropdown={toggleDropdownMock}
      />
    );

    const dropdownButton = getByTestId('drop-down');
    fireEvent.click(dropdownButton);

    // Assertions
    expect(dropdownButton).toBeTruthy();
    expect(toggleDropdownMock).toHaveBeenCalled();
  });
});
