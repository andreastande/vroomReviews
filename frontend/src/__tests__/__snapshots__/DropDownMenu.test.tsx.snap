// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown-button-wrapper"
  >
    <button
      class="dropdown-button"
      data-testid="drop-down"
    >
      <label
        class="ddb-label"
      />
      <i
        class="dropdown-arrow"
      />
    </button>
    <div
      class="dropdown closed"
    >
      <menu
        class="options-wrapper"
      >
        <div
          class="dropdown-button-inside-wrapper"
        >
          <button
            class="dropdown-button-inside"
          >
            <span>
              Option1
            </span>
          </button>
        </div>
        <div
          class="dropdown-button-inside-wrapper"
        >
          <button
            class="dropdown-button-inside"
          >
            <span>
              Option2
            </span>
          </button>
        </div>
        <div
          class="dropdown-button-inside-wrapper"
        >
          <button
            class="dropdown-button-inside"
          >
            <span>
              Option3
            </span>
          </button>
        </div>
      </menu>
      <div
        class="arrow-wrapper"
        style="justify-content: center;"
      />
    </div>
    <button
      class="clear-button"
    >
      <img
        src="/project2/src/assets/images/clear-icon.png"
      />
    </button>
  </div>
</DocumentFragment>
`;
